PEP 622: Ditch leading dots for name loads (#1499)
python/peps: pep-0622.rst

This is used to match against constants and enum values.
Every dotted name in a pattern is looked up using normal Python name
resolution rules, and the value is used for comparison by equality with
the match subject (same as for literals)::

  from enum import Enum

  class Sides(str, Enum):
      SPAM = "spam"
      EGGS = "eggs"
      ...

  match entree[-1]:
      case Sides.SPAM:  # Compares entree[-1] == Sides.SPAM.

>           response = f"Have you got anything without {Sides.SPAM}?"
>       case side:  # Assigns side == entree[-1].
>           response = f"Well, could I have their {Sides.SPAM} instead of the {side} then?"

Reviewer: I'd suggest not distracting the reader with `{Sides.SPAM}` in the f-string.
```suggestion
          response = "Have you got anything without Spam?"
      case side:  # Assigns side == entree[-1].
          response = f"Well, could I have their Spam instead of the {side} then?"
```

Note that it is not possible for unqualified names (such as ``side``

> in the example above) to be used as constant value patterns.  See

Reviewer: Somehow using `side` above feels distracting too. The original phrase was a bit awkward already. How about this?
```suggestion
Note that there is no way to use unqualified names as constant value
patterns (they always denote variables to be captured).  See
```

This is used to match against constants and enum values.
Every dotted name in a pattern is looked up using normal Python name
resolution rules, and the value is used for comparison by equality with

> the matching expression (same as for literals)::

Reviewer: ```suggestion
the match subject (same as for literals)::
```

* Use some implicit rules. For example if a name was defined in the global
  scope, then it refers to a constant, rather than represents a capture pattern::

    assert "spam" in globals() and "spam" not in locals()

>     assert "side" not in globals()

Reviewer: Maybe just state the conditions in English? I totally misread these the first one I saw them.

  to be assigned to.  An earlier version of this proposal used a
  leading-dot rule::

    match entree[-1]:
        case .spam: ...  # Compares entree[-1] == spam.
        case side: ...   # Assigns side = entree[-1].

  While potentially useful, it introduces strange-looking new syntax
  without making the pattern syntax any more expressive.  Indeed,
  named constants can be made to work with the existing rules by
  converting them to ``Enum`` types, or enclosing them in their own
  namespace (considered by the authors to be one honking great idea)::

    match entree[-1]:
        case Sides.SPAM: ...  # Compares entree[-1] == Sides.SPAM.
        case side: ...        # Assigns side = entree[-1].

  If needed, the leading-dot rule (or a similar variant) could be
  added back later with no backward-compatibility issues.

* There was also on idea to make lookup semantics the default, and require
  ``$`` or ``?`` to be used in capture patterns::

    match entree[-1]:
        case spam: ...   # Compares entree[-1] == spam.
        case side?: ...  # Assigns side = entree[-1].

  But capture patterns are more common in typical code, so it is

>   undesirable to require special syntax for them.

Reviewer: There are more arguments against marking capture variables with sigils:
- None of the other languages surveyed do this (and there's no precedent in Python either).
- It would look confusing to see
  ```
  case (%x, %y):
      return Point(x, y)  # Why not Point(%x, %y)
  ```

  disambiguate grouping in patterns, e.g. in ``Point(x, y=(y :=
  complex()))``.

* Introduce a special symbol, for example ``.``, ``$``, or ``^`` to
  indicate that a given name is a constant to be matched against, not
  to be assigned to.  An earlier version of this proposal used a
  leading-dot rule::

    FOO = 1
    value = 0

>     match value:

Reviewer: This whole example is confusing.

        case .FOO:  # This would not be matched
            ...
        case BAR:  # This would be matched
            ...

  While potentially useful, it introduces strange-looking new syntax
  without making the pattern syntax any more expressive.  Indeed,
  named constants can be made to work with the existing rules by
  converting them to ``Enum`` types, or enclosing them in their own
  namespace (considered by the authors to be one honking great idea)::

    class Consts:
        FOO = 1
        ...

    value = 0

>     match value:

Reviewer: This too
